src/lightning/data/streaming/dataset_optimizer.py

```diff
@@ -790,6 +819,14 @@ def _upload_index(self, cache_dir: str, num_nodes: int, node_rank: Optional[int]
                 node_index_filepath = os.path.join(cache_dir, os.path.basename(remote_filepath))
                 if obj.scheme == "s3":
                     obj = parse.urlparse(remote_filepath)
+                    # Wait for the index to exist on s3.
+                    while True:
+                        try:
+                            s3.head_object(Bucket=obj.netloc, Key=obj.path.lstrip("/"))
+                            break
+                        except Exception as e:
+                            print(e)
+                            pass
```

Comment:
Maybe worth adding a small timeout here & even some backoff if it's a 503 slow down error

Reply:
Totally. I thought of it while taking a shower. But you had the review already ;)